crates/sui-core/src/validator_tx_finalizer.rs

```diff
@@ -170,6 +199,45 @@ where
         self.metrics.finalization_succeeded(start_time);
         Ok(true)
     }
+
+    // We want to avoid all validators waking up at the same time to finalize the same transaction.
+    // That can lead to a waste of resource and flood the network unnecessarily.
+    // Here we use the transaction digest to determine an order of all validators.
+    // The first few will wake up one by one with incremental delays to finalize the transaction.
+    // The rest will wake up at the same time afterward. The hope is that the first few should be
+    // able to finalize the transaction, and the rest will see it already executed and do not need
+    // to do anything.
+    fn determine_finalization_delay(
+        &self,
+        tx_digest: &TransactionDigest,
+    ) -> anyhow::Result<(usize, Duration)> {
+        // the 32 is as requirement of the default StdRng::from_seed choice
+        let digest_bytes = tx_digest.into_inner();
+
+        // permute the validators deterministically, based on the digest
+        let mut rng = StdRng::from_seed(digest_bytes);
+        let order = self
+            .agg
+            .committee
+            .shuffle_by_stake_with_rng(None, None, &mut rng);
```

Comment:
i'm guessing you copied this code from `order_validators_for_submission` in consensus_adapter - is it worth trying to move that function to a common location?